sherpa-onnx/csrc/funasr-nano-tokenizer.cc

```diff
@@ -0,0 +1,1366 @@
+                                       const std::string &b) {
+  std::string k = a;
+  k.push_back('\t');
+  k.append(b);
+  return k;
+}
+
+}  // namespace
+
+// Parse tokenizer.json added_tokens: extract objects with {id, content, ...}
+bool ParseAddedTokensFromTokenizerJson(
+    const std::string &blob, std::vector<FunASRNanoTokenizer::AddedToken> *out) {
+  if (!out) return false;
+  out->clear();
+
+  JsonReader r(blob);
+  if (!r.SeekToKey("added_tokens")) return true;
+  if (!r.Consume(':')) return false;
+  if (!r.Consume('[')) return false;
+
+  r.SkipWs();
+  if (r.Consume(']')) return true;
+
+  while (true) {
+    if (!r.Consume('{')) return false;
+    FunASRNanoTokenizer::AddedToken t;
+
+    r.SkipWs();
+    if (!r.Consume('}')) {
+      while (true) {
+        std::string k;
+        if (!r.ParseString(&k)) return false;
+        if (!r.Consume(':')) return false;
+
+        if (k == "id") {
+          int64_t v = 0;
+          if (!r.ParseInt64(&v)) return false;
+          t.id = static_cast<int32_t>(v);
+        } else if (k == "content") {
+          if (!r.ParseString(&t.content)) return false;
+        } else if (k == "single_word") {
+          if (!r.ParseBool(&t.single_word)) return false;
+        } else if (k == "lstrip") {
+          if (!r.ParseBool(&t.lstrip)) return false;
+        } else if (k == "rstrip") {
+          if (!r.ParseBool(&t.rstrip)) return false;
+        } else if (k == "normalized") {
+          if (!r.ParseBool(&t.normalized)) return false;
+        } else if (k == "special") {
+          if (!r.ParseBool(&t.special)) return false;
+        } else {
+          if (!r.SkipValue()) return false;
+        }
+
+        r.SkipWs();
+        if (r.Consume('}')) break;
+        if (!r.Consume(',')) return false;
+      }
+    }
+
+    if (t.id >= 0 && !t.content.empty()) {
+      out->push_back(std::move(t));
+    }
+
+    r.SkipWs();
+    if (r.Consume(']')) return true;
+    if (!r.Consume(',')) return false;
+  }
+}
+
+// Build trie for AddedTokens longest match (byte-wise).
+void BuildAddedTokensTrie(
+    const std::vector<FunASRNanoTokenizer::AddedToken> &tokens,
+    std::vector<FunASRNanoTokenizer::TrieNode> *trie) {
+  if (!trie) return;
+  trie->clear();
+  trie->push_back(FunASRNanoTokenizer::TrieNode{});
+  for (int32_t i = 0; i < static_cast<int32_t>(tokens.size()); ++i) {
+    const auto &tok = tokens[i];
+    int32_t node = 0;
+    for (uint8_t b : std::vector<uint8_t>(tok.content.begin(),
+                                          tok.content.end())) {
+      auto it = (*trie)[node].next.find(b);
+      if (it == (*trie)[node].next.end()) {
+        int32_t new_node = static_cast<int32_t>(trie->size());
+        trie->push_back(FunASRNanoTokenizer::TrieNode{});
+        (*trie)[node].next.emplace(b, new_node);
+        node = new_node;
+      } else {
+        node = it->second;
+      }
+    }
+    (*trie)[node].token_index = i;
+  }
+}
+
+void MergeVocabAndAddedTokens(
```

Comment:
```suggestion
static void MergeVocabAndAddedTokens(
```

Please make functions used only in this file `static`.